psamm/moma.py

```diff
@@ -79,84 +100,332 @@ def __init__(self, model, solver):
+        with self.constraints(wt_biomass == wt_obj):
+            self._prob.set_objective(self._v_wt.sum(self._all_reactions()))
+            result = self._solve(lp.ObjectiveSense.Minimize)
+        return result
+
+    # Solves for the FBA biomass. Uses the result returned by _solve_fba and
+    # finds the value of the objective (biomass) function.
+    def _get_fba_biomass(self, objective):
+        """Returns the flux for the max biomass found by FBA."""
+        # Run FBA and store the result
+        flux_result = self._solve_fba(objective)
+        # Pull out the biomass value
+        return flux_result.get_value(self._v_wt(objective))
+
+    def minimize_lp2(self, objective, wt_fluxes):
+        """Implemetnation of the lin_MOMA algorithm.
+
+        We try to maximize biomass, by keeping the fluxes relitively the same
+        as the wildtype. This helps avoid the assumption that an organism will
+        perform optimally directly after removing a gene.
+
+        This might not be the optimal minimization solution for the wild type
+        and the knockout vectors.
+
+        Minimizes sum of |wild type - knockout|
+
+        Args:
+            objective: Biomass reaction for the model.
+            wt_fluxes: All the wt_fluxes found by FBA in the form of a
+                dictionary. Use :meth: _get_fba_flux(objective) to return
+                a dictionary.
+        """
+        # These are all of our non eachange reactions
+        reactions = set(self._adjustment_reactions())
+
+        # First times in the loop
         if self._z is None:
             self._z = z = self._prob.namespace()
-            reactions = set(self._adjustment_reactions())
             z.define(reactions, lower=0)
 
-            zs = z.set(reactions)
-            vs_wt = self._v_wt.set(reactions)
-            vs = self._v.set(reactions)
+        # After the first time, we can just use the same z object as before
+        else:
+            z = self._z
+
+        # Get the v model and all the reactions 'variables' associated with it
+        v = self._v
+
+        # Create a constraints object for this model
+        constr = self.constraints()
+
+        # Loop through all the flux reactions
+        for f_reaction, f_value in iteritems(wt_fluxes):
+            # Makes sure the reaction we are trying to put a constraint on is
+            # a non exchange reaction
+            if f_reaction in reactions:
+                # Add the constraint that finds the optimal solution, such
+                # that the difference between the wildtype flux is similar
+                # to the knockout flux.
+                constr.add(
+                    z(f_reaction) >= f_value - v(f_reaction),
+                    f_value - v(f_reaction) >= -z(f_reaction))
+
+        # If we minimize the sum of the z vector then we will minimize
+        # the |vs_wt - vs| from above
+        self._prob.set_objective(z.sum(self._adjustment_reactions()))
```

Comment:
You can use the `reactions` variable here instead of calling `self._adjustment_reactions()` again.